### PR TITLE
crimson: make backtraces more human-readable with boost::stacktrace and addr2line

### PR DESCRIPTION
**Pull request by rzarzynski**
This frees a developer from the need to pass the backtrace through `seastar-addr2line`. After the change:

```
[rzarzynski@o06 build]$ ./bin/unittest-seastar-errorator --smp 1 
[==========] Running 2 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 2 tests from errorator_test_t
[ RUN      ] errorator_test_t.basic
Aborting on shard 0.
Backtrace:
 0# print_with_backtrace(std::basic_string_view<char, std::char_traits<char> >) at /home/rzarzynski/ceph1/src/crimson/common/fatal_signal.cc:59
 1# FatalSignal::signaled(int) at /home/rzarzynski/ceph1/src/crimson/common/fatal_signal.cc:78
 2# FatalSignal::install_oneshot_signal_handler<11>()::{lambda(int, siginfo_t*, void*)#1}::_FUN(int, siginfo_t*, void*) at /home/rzarzynski/ceph1/src/crimson/common/fatal_signal.cc:41
 3# 0x00007F5640897B30 in /lib64/libpthread.so.0
 4# auto crimson::do_until<errorator_test_t::test_do_until()::{lambda()#1}>(errorator_test_t::test_do_until()::{lambda()#1}) at /home/rzarzynski/ceph1/src/test/crimson/test_errorator.cc:21
 5# seastar::noncopyable_function<void ()>::direct_vtable_for<seastar::async<errorator_test_t_basic_Test::TestBody()::{lambda()#1}>(seastar::thread_attributes, std::decay&&, (std::decay<errorator_test_t_basic_Test::TestBody()::{lambda()#1}>::type&&)...)::{lambda()#1}>::call(seastar::noncopyable_function<void ()> const*) at /home/rzarzynski/ceph1/src/seastar/include/seastar/core/future.hh:1473
 6# seastar::thread_context::main() at /home/rzarzynski/ceph1/src/seastar/include/seastar/core/future.hh:924
Segmentation fault (core dumped)
```

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by rzarzynski** [comment]
Hmm, perhaps we should say `Requires: binutils` in `ceph.spec.in`.

File: src/crimson/common/fatal_signal.cc [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+void FatalSignal::install_oneshot_signals_handler()
+{
+  (install_oneshot_signal_handler<SigNums>() , ...);
+}
+
+template <int SigNum>
+void FatalSignal::install_oneshot_signal_handler()
+{
+  struct sigaction sa;
+  sa.sa_sigaction = [](int sig, siginfo_t *info, void *p) {
```
Comment (rzarzynski) [inline src/crimson/common/fatal_signal.cc:36]: FOLLOW-UP: let's make a use of `info` and dump `si_addr` for `SIGSEGV` as it's really, really useful to know the address which caused the problem.

File: src/common/BackTrace.cc [hunk 2]
```diff
@@ -2,29 +2,33 @@
...
+BackTrace::BackTrace(std::size_t s)
+  : bt{s, max}
+{}
+
 void BackTrace::print(std::ostream& out) const
 {
   out << " " << pretty_version_to_str() << std::endl;
-  for (size_t i = skip; i < size; i++) {
-    out << " " << (i-skip+1) << ": " << demangle(strings[i]) << std::endl;
+  for (size_t i = 0; i < bt.size(); i++) {
```
Comment (tchaikov) [inline src/common/BackTrace.cc:22]: why is `skip` skipped?

File: src/common/BackTrace.cc [hunk 3]
```diff
@@ -2,29 +2,33 @@
 // vim: ts=8 sw=2 smarttab
 
 #include <ostream>
-#include <cxxabi.h>
 #include <string.h>
 
+#include <boost/stacktrace.hpp>
```
Comment (tchaikov) [inline src/common/BackTrace.cc:7]: i'd suggest move the `crommon/BackTrace.cc` change out into a separated PR, so we can have this change merged sooner. also, AFAIR, the output format of demangled symbol of boost.backtrace is a little bit different from what we have now. not sure if this matters.

Comment (tchaikov) [reply src/common/BackTrace.cc:7]:
```
[ RUN      ] BackTrace.Basic
../src/test/common/test_back_trace.cc:43: Failure
Value of: std::regex_match(lines[lineno], e)
  Actual: false
Expected: true
[  FAILED  ] BackTrace.Basic (2 ms)
```

indeed.

Comment (jdurgin) [reply src/common/BackTrace.cc:7]: the format does matter for crash-reports that telemetry analyses - @yaarith fyi

File: src/common/BackTrace.cc (hunk 2 above)
Comment (rzarzynski) [reply src/common/BackTrace.cc:22]: I think it's just delegated outside, to the `stacktrace`. Before the change `skip` was set basing the value passed to the constructor:

```diff
-   explicit BackTrace(int s) : skip(s) {
```

which is now supplied to `bt`

```diff
+BackTrace::BackTrace(int s)
+  : bt{s, max}
+{}
```

File: src/common/BackTrace.cc (hunk 3 above)
Comment (rzarzynski) [reply src/common/BackTrace.cc]: Sure, dissected into https://github.com/ceph/ceph/pull/40947.

**Comment by tchaikov** [comment]
jenkins test docs

**Review by tchaikov — APPROVED** [review]

File: src/crimson/common/fatal_signal.cc [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    FatalSignal::signaled(sig);
+    ::signal(sig, SIG_DFL);
+  };
+  sigfillset(&sa.sa_mask);
+  sa.sa_flags = SA_SIGINFO | SA_RESTART;
+  if constexpr (SigNum == SIGSEGV) {
+    sa.sa_flags |= SA_ONSTACK;
+  }
+  [[maybe_unused]] auto r = ::sigaction(SigNum, &sa, nullptr);
+  assert(r);
```
Comment (athanatos) [inline src/crimson/common/fatal_signal.cc:49]: @rzarzynski Is this actually correct?  The man page seems to say that 0 means success?

Comment (rzarzynski) [reply src/crimson/common/fatal_signal.cc:49]: Oops, I missed the negation here. The problem slipped in undetected b/c using productions builds.
Fixing. Thanks for pointing this out!

Comment (rzarzynski) [reply src/crimson/common/fatal_signal.cc:49]: https://github.com/ceph/ceph/pull/41010

Comment (tchaikov) [reply src/crimson/common/fatal_signal.cc:49]: thank you for your prompt fix. should have been more carefully when reviewing this change. =(

**Pull request merged**